robot/packages/utils.go

```diff
@@ -302,3 +287,87 @@ func commonCleanup(logger logging.Logger, expectedPackageDirectories map[string]
 	}
 	return allErrors
 }
+
+type syncStatus string
+
+const (
+	syncStatusDownloading syncStatus = "downloading"
+	syncStatusUnpacking   syncStatus = "unpacking"
+	syncStatusDone        syncStatus = "done"
+)
+
+type packageSyncFile struct {
+	PackageID       string     `json:"package_id"`
+	Version         string     `json:"version"`
+	ModifiedTime    time.Time  `json:"modified_time"`
+	Status          syncStatus `json:"sync_status"`
+	TarballChecksum string     `json:"tarball_checksum"`
+}
+
+var statusFileExt = ".status.json"
+
+func packageIsSynced(pkg config.PackageConfig, packagesDir string, logger logging.Logger) bool {
+	syncFile, err := readStatusFile(pkg, packagesDir)
+	if err != nil {
+		utils.UncheckedError(err)
+	}
+	if syncFile.PackageID == pkg.Package && syncFile.Version == pkg.Version && syncFile.Status == syncStatusDone {
+		logger.Debugf("Package already downloaded at %s, skipping.", pkg.LocalDataDirectory(packagesDir))
+		return true
+	}
+	logger.Debugf("Package is not in sync for %s: status of '%s' (file) != '%s' (expected) and version of '%s' (file) != '%s' (expected)",
+		pkg.Package, syncFile.Status, syncStatusDone, syncFile.Version, pkg.Version)
+	return false
+}
+
+func packagesAreSynced(packages []config.PackageConfig, packagesDir string, logger logging.Logger) bool {
+	for _, pkg := range packages {
+		if !packageIsSynced(pkg, packagesDir, logger) {
+			return false
+		}
+	}
+	return true
+}
+
+func getSyncFileName(packageLocalDataDirectory string) string {
+	return packageLocalDataDirectory + statusFileExt
```

Comment:
not sure this is a legit concern, but is there a case where packageLocalDataDirectory has a trailing slash? if yes consider adding [filepath.Clean](https://pkg.go.dev/path/filepath#Clean)

this may be a nit not worth doing